lib/deps/request-browser.js

```diff
@@ -3,7 +3,111 @@
 var createBlob = require('./blob.js');
 var utils = require('../utils');
 
-module.exports = function(options, callback) {
+function wrappedFetch() {
+  var args = Array.prototype.slice.call(arguments); // arguments to array
+
+  var wrappedPromise = {};
+
+  var promise = new global.Promise(function(resolve, reject) {
+    wrappedPromise.resolve = resolve;
+    wrappedPromise.reject = reject;
+  });
+
+  wrappedPromise.then = promise.then.bind(promise);
+  wrappedPromise.catch = promise.catch.bind(promise);
+  wrappedPromise.promise = promise;
+
+  global.fetch.apply(null, args).then(function(response) {
+    wrappedPromise.resolve(response);
+  }, function(error) {
+    wrappedPromise.reject(error);
+  }).catch(function(error) {
+    wrappedPromise.catch(error);
+  });
+
+  return wrappedPromise;
+}
+
+function fetchRequest(options, callback) {
+  var wrappedPromise, timer, fetchResponse;
+
+  var headers = new global.Headers();
+
+  var fetchOptions = {
+    method: options.method,
+    credentials: 'include',
+    headers: headers
+  };
+
+  if (options.json) {
+    headers.set('Accept', 'application/json');
+    headers.set('Content-Type', options.headers['Content-Type'] ||
+      'application/json');
+  }
+
+  if (options.body && (options.body instanceof Blob)) {
+    utils.readAsBinaryString(options.body, function(binary) {
+      fetchOptions.body = utils.fixBinary(binary);
+    });
+  } else if (options.body &&
+             options.processData &&
+             typeof options.body !== 'string') {
+    fetchOptions.body = JSON.stringify(options.body);
+  } else if ('body' in options) {
+    fetchOptions.body = options.body;
+  } else {
+    fetchOptions.body = null;
+  }
+
+  Object.keys(options.headers).forEach(function(key) {
+    if (options.headers.hasOwnProperty(key)) {
+      headers.set(key, options.headers[key]);
+    }
+  });
+
+  wrappedPromise = wrappedFetch(options.url, fetchOptions);
+
+  if (options.timeout > 0) {
+    timer = setTimeout(function() {
+      wrappedPromise.reject(new Error('Load timeout for resource: ' +
+        options.url));
+    }, options.timeout);
+  }
+
+  wrappedPromise.promise.then(function(response) {
+    var result;
+
+    fetchResponse = response;
+
+    if (options.timeout > 0) {
+      clearTimeout(timer);
+    }
+
+    if (response.status >= 200 && response.status < 300) {
+      if (options.binary) {
+        result = response.blob();
+      } else {
+        result = response.text();
+      }
+    } else {
+      result = response.json();
+    }
+
+    return result;
```

Comment:
Generally lean towards early returns and it looks like it would be cleaner?

   if (response.status >= 200 && response.status < 300) { 
     return options.binary ? response.blob() : response.text();
   } 
   return response.json;